`pytest.warns()` asserts when `pytest.skip()` is called

When `pytest.skip()` is used inside `pytest.warns()` (and no warning is emitted), the test fails. I'm not sure if this is intentional but it feels a bit off, given that the test would normally be skipped instead. I've hit this in the wild in [urllib3](https://github.com/urllib3/urllib3/)'s test suite.

Quick reproducer:

```python
import pytest


def test_foo():
    with pytest.warns(UserWarning):
        pytest.skip("skipping")
```

Yields:

```pytb
$ python -m pytest
========================================================= test session starts =========================================================
platform linux -- Python 3.11.7, pytest-8.1.0.dev116+g4546d5445, pluggy-1.4.0
rootdir: /tmp/test
collected 1 item                                                                                                                      

test_foo.py F                                                                                                                   [100%]

============================================================== FAILURES ===============================================================
______________________________________________________________ test_foo _______________________________________________________________

    def test_foo():
        with pytest.warns(UserWarning):
>           pytest.skip("skipping")
E           Skipped: skipping

test_foo.py:6: Skipped

During handling of the above exception, another exception occurred:

    def test_foo():
>       with pytest.warns(UserWarning):
E       Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
E        Emitted warnings: [].

test_foo.py:5: Failed
======================================================= short test summary info =======================================================
FAILED test_foo.py::test_foo - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
========================================================== 1 failed in 0.02s ==========================================================
```

pip list:

```
Package    Version
---------- -----------------------
iniconfig  2.0.0
packaging  23.2
pip        23.3.2
pluggy     1.4.0
pytest     8.1.0.dev116+g4546d5445
setuptools 69.0.3
```

------

- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

--- a/changelog/11907.bugfix.rst
+++ b/changelog/11907.bugfix.rst
@@ -1,0 +1,1 @@
+Fix a regression in pytest 8.0.0 whereby calling :func:`pytest.skip` and similar control-flow exceptions within a :func:`pytest.warns()` block would get suppressed instead of propagating.

--- a/src/_pytest/recwarn.py
+++ b/src/_pytest/recwarn.py
@@ -20,6 +20,7 @@ import warnings
 
 from _pytest.deprecated import check_ispytest
 from _pytest.fixtures import fixture
+from _pytest.outcomes import Exit
 from _pytest.outcomes import fail
 
 
@@ -301,6 +302,17 @@ class WarningsChecker(WarningsRecorder):
         super().__exit__(exc_type, exc_val, exc_tb)
 
         __tracebackhide__ = True
+
+        # BaseExceptions like pytest.{skip,fail,xfail,exit} or Ctrl-C within
+        # pytest.warns should *not* trigger "DID NOT WARN" and get suppressed
+        # when the warning doesn't happen. Control-flow exceptions should always
+        # propagate.
+        if exc_val is not None and (
+            not isinstance(exc_val, Exception)
+            # Exit is an Exception, not a BaseException, for some reason.
+            or isinstance(exc_val, Exit)
+        ):
+            return
 
         def found_str():
             return pformat([record.message for record in self], indent=2)

--- a/testing/test_recwarn.py
+++ b/testing/test_recwarn.py
@@ -5,9 +5,10 @@ from typing import Optional
 from typing import Type
 import warnings
 
-from _pytest.pytester import Pytester
-from _pytest.recwarn import WarningsRecorder
 import pytest
+from pytest import ExitCode
+from pytest import Pytester
+from pytest import WarningsRecorder
 
 
 def test_recwarn_stacklevel(recwarn: WarningsRecorder) -> None:
@@ -478,6 +479,71 @@ class TestWarns:
             with pytest.raises(ValueError, match="some exception"):
                 warnings.warn("some warning", category=FutureWarning)
                 raise ValueError("some exception")
+
+    def test_skip_within_warns(self, pytester: Pytester) -> None:
+        """Regression test for #11907."""
+        pytester.makepyfile(
+            """
+            import pytest
+
+            def test_it():
+                with pytest.warns(Warning):
+                    pytest.skip("this is OK")
+            """,
+        )
+
+        result = pytester.runpytest()
+        assert result.ret == ExitCode.OK
+        result.assert_outcomes(skipped=1)
+
+    def test_fail_within_warns(self, pytester: Pytester) -> None:
+        """Regression test for #11907."""
+        pytester.makepyfile(
+            """
+            import pytest
+
+            def test_it():
+                with pytest.warns(Warning):
+                    pytest.fail("BOOM")
+            """,
+        )
+
+        result = pytester.runpytest()
+        assert result.ret == ExitCode.TESTS_FAILED
+        result.assert_outcomes(failed=1)
+        assert "DID NOT WARN" not in str(result.stdout)
+
+    def test_exit_within_warns(self, pytester: Pytester) -> None:
+        """Regression test for #11907."""
+        pytester.makepyfile(
+            """
+            import pytest
+
+            def test_it():
+                with pytest.warns(Warning):
+                    pytest.exit()
+            """,
+        )
+
+        result = pytester.runpytest()
+        assert result.ret == ExitCode.INTERRUPTED
+        result.assert_outcomes()
+
+    def test_keyboard_interrupt_within_warns(self, pytester: Pytester) -> None:
+        """Regression test for #11907."""
+        pytester.makepyfile(
+            """
+            import pytest
+
+            def test_it():
+                with pytest.warns(Warning):
+                    raise KeyboardInterrupt()
+            """,
+        )
+
+        result = pytester.runpytest_subprocess()
+        assert result.ret == ExitCode.INTERRUPTED
+        result.assert_outcomes()
 
 
 def test_raise_type_error_on_non_string_warning() -> None:

